lib/ASTGen/Sources/ASTGen/Attrs.swift

```diff
@@ -0,0 +1,1132 @@
+    args: inout Slice<LabeledExprListSyntax>,
+    label: SyntaxText? = nil
+  ) -> BridgedStringRef? {
+    return self.generateConsumingAttrOption(args: &args, label: label) {
+      self.generateStringLiteralTextIfNotInterpolated(expr: $0)
+    }
+  }
+
+  /// Extracts single identifier token argument from the `attribute`, calls the
+  /// specified function with the token, then returns the result. If `valueIfOmitted`
+  /// is not nil, and there's no argument clause in the `attribute`, this returns
+  /// the `valueIfOmitted` value.
+  /// Returns `nil` if the argument was not a single token, the callback returned `nil`,
+  /// or there was no argument clause and `valueIfOmitted` is not specified. In such cases,
+  /// this emits diagnostics.
+  func generateSingleAttrOption<Result>(
+    attribute node: AttributeSyntax,
+    _ valueGeneratorFunction: (TokenSyntax) -> Result?,
+    valueIfOmitted: Result? = nil
+  ) -> Result? {
+    guard node.leftParen != nil, let arguments = node.arguments else {
+      if let valueIfOmitted {
+        return valueIfOmitted
+      }
+      // TODO: Diagnose.
+      return nil
+    }
+
+    guard var arguments = arguments.as(LabeledExprListSyntax.self)?[...] else {
+      // TODO: Diagnose.
+      return nil
+    }
+
+    let result = self.generateConsumingPlainIdentifierAttrOption(
+      args: &arguments,
+      valueGeneratorFunction
+    )
+    guard let result else {
+      return nil
+    }
+
+    if !arguments.isEmpty {
+      // TODO: Diagnose.
+      // But still can return the value.
+    }
+
+    return result
+  }
+}
+
+// MARK: - Decl modifiers
+extension ASTGenVisitor {
+  func generate(declModifier node: DeclModifierSyntax) -> BridgedDeclAttribute? {
+    switch node.name.keywordKind {
+    case .private:
+      return self.generateAccessControlAttr(declModifier: node, level: .private)
+    case .fileprivate:
+      return self.generateAccessControlAttr(declModifier: node, level: .filePrivate)
+    case .internal:
+      return self.generateAccessControlAttr(declModifier: node, level: .internal)
+    case .package:
+      return self.generateAccessControlAttr(declModifier: node, level: .package)
+    case .public:
+      return self.generateAccessControlAttr(declModifier: node, level: .public)
+    case .open:
+      return self.generateAccessControlAttr(declModifier: node, level: .open)
+    case .weak, .unowned:
+      return self.generateReferenceOwnershipAttr(declModifier: node)?.asDeclAttribute
+    default:
+      // Other modifiers are all "simple" attributes.
+      let kind = BridgedDeclAttrKind(from: node.name.rawText.bridged)
+      guard kind != .none else {
+        // TODO: Diagnose?
+        assertionFailure("unknown decl modifier")
+        return nil
+      }
+      return self.generateSimpleDeclAttr(declModifier: node, kind: kind)
+    }
+  }
+
+  func generateAccessControlAttr(declModifier node: DeclModifierSyntax, level: BridgedAccessLevel)
+    -> BridgedDeclAttribute?
```

Comment:
FWIW I’m not a huge fan of this formatting generated by swift-format and usually manually change it to break the arguments instead.